### PR TITLE
reverting the sort logic in favour of spliting and joining the collec…

### PR DESCRIPTION
**Pull request by cheemau**
…tion to retain the message order, also checking the schema type is avro before saving avro.schema attribute value to prevent json or protbuf dont write this attribute

### CODE REVIEW

File: nifi-pulsar-processors/src/main/java/org/apache/nifi/processors/pulsar/pubsub/ConsumePulsarRecord.java [hunk 1]
```diff
@@ -195,240 +197,243 @@ private List<Message<GenericRecord>> getMessages(final Consumer<GenericRecord> c
...
+
+                // if there's no record set actively being written, begin one
+                byte[] data = msg.getData();
+                if (lastMessage == null) {
+                    flowFile = session.create();
+                    flowFile = session.putAllAttributes(flowFile, currentAttributes);
+                    if (msg.getReaderSchema().isPresent() && msg.getReaderSchema().get().getSchemaInfo().getType() == SchemaType.AVRO) {
+                        String msgSchema = new String(msg.getReaderSchema().get().getSchemaInfo().getSchema());
+                        flowFile = session.putAttribute(flowFile, "avro.schema", msgSchema);
+                        schema = new SimpleRecordSchema(
```
Comment (z-kovacs) [inline nifi-pulsar-processors/src/main/java/org/apache/nifi/processors/pulsar/pubsub/ConsumePulsarRecord.java]: we have to replace the schema effectively here? `schema = getSchema(flowFile, readerFactory, data);` will not do the right thing?

**Comment by z-kovacs** [comment]
please add the tests we discussed

**Review by z-kovacs — CHANGES_REQUESTED** [review]
pls add the tests dicsussed

File: nifi-pulsar-processors/src/main/java/org/apache/nifi/processors/pulsar/pubsub/ConsumePulsarRecord.java [hunk 2]
```diff
@@ -195,240 +197,238 @@ private List<Message<GenericRecord>> getMessages(final Consumer<GenericRecord> c
...
+        if (async) {
+            getAckService().submit(new Callable<Object>() {
+                @Override
+                public Object call() throws Exception {
+                    return consumer.acknowledgeAsync(msg).get();
+                }
+            });
+        } else {
+            consumer.acknowledge(msg);
+            ;
```
Comment (z-kovacs) [inline nifi-pulsar-processors/src/main/java/org/apache/nifi/processors/pulsar/pubsub/ConsumePulsarRecord.java]: extra `;`

File: nifi-pulsar-processors/src/test/java/org/apache/nifi/processors/pulsar/pubsub/async/TestAsyncConsumePulsarRecord.java [hunk 3]
```diff
@@ -166,4 +247,27 @@ public void mappedAttributesTest() throws PulsarClientException {
...
+                                                            String schemaString,
+                                                            SchemaType schemaType
+    ) {
+        Message mockA = mock(Message.class);
+        Schema<Object> schema = mock(Schema.class);
+        SchemaInfo schemaInfo = mock(SchemaInfo.class);
+        when(mockA.getReaderSchema()).thenReturn(Optional.of(schema));
+        when(schema.getSchemaInfo()).thenReturn(schemaInfo);
+        when(schemaInfo.getType()).thenReturn(schemaType);
+        when(schemaInfo.getSchema()).thenReturn(schemaString.getBytes());
```
Comment (z-kovacs) [inline nifi-pulsar-processors/src/test/java/org/apache/nifi/processors/pulsar/pubsub/async/TestAsyncConsumePulsarRecord.java:262]: are we certain that this happens when you run in real life?

File: nifi-pulsar-processors/src/test/java/org/apache/nifi/processors/pulsar/pubsub/async/TestAsyncConsumePulsarRecord.java [hunk 4]
```diff
@@ -91,36 +104,104 @@ public void singleMessageMultiRecordsTest() throws PulsarClientException {
...
 
+    /*
+     * Send multiple messages on different topics,
+     * check if it creates two flow files by retaining the order of messages
+     */
+    @Test
+    public void multipleGoodMessagesOnTwoTopicsCreatesMultipleRecordsTest() throws IOException {
+
+        List<Message<GenericRecord>> mockMessages = new ArrayList<>();
+        mockMessages.add(createTestMessage("A,9".getBytes(), null, singletonMap("prop", "val"), DEFAULT_TOPIC, "", SchemaType.JSON));
```
Comment (z-kovacs) [inline nifi-pulsar-processors/src/test/java/org/apache/nifi/processors/pulsar/pubsub/async/TestAsyncConsumePulsarRecord.java:113]: this is a bit disturbing - you are sending in random binary stating it's json and it does not blow up.

File: nifi-pulsar-processors/src/test/java/org/apache/nifi/processors/pulsar/pubsub/async/TestAsyncConsumePulsarRecord.java [hunk 5]
```diff
@@ -91,36 +104,104 @@ public void singleMessageMultiRecordsTest() throws PulsarClientException {
...
+     * Send multiple messages on different topics while updating schema for one topic,
+     * check if it creates two three files by retaining the order of messages
+     */
+    @Test
+    public void multipleGoodMessagesWithSchemaUpdateOnTwoTopicsTest() throws IOException {
+        String schema1 = "{\"type\": \"record\",\"name\": \"ExampleRecord\",\"fields\": [{\"name\": \"field1\", \"type\": \"int\"}]}\r\n";
+        String schema1_updated = "{\"type\": \"record\",\"name\": \"ExampleRecord\",\"fields\": [{\"name\": \"field1\", \"type\": \"int\"},{\"name\": \"field2\", \"type\": [\"null\", \"string\"], \"default\": null}]}\r\n";
+
+        List<Message<GenericRecord>> mockMessages = new ArrayList<>();
+        mockMessages.add(createTestMessage("A,9".getBytes(), null, singletonMap("prop", "val"), DEFAULT_TOPIC, schema1, SchemaType.AVRO));
```
Comment (z-kovacs) [inline nifi-pulsar-processors/src/test/java/org/apache/nifi/processors/pulsar/pubsub/async/TestAsyncConsumePulsarRecord.java:145]: also very disturbing

File: nifi-pulsar-processors/src/test/java/org/apache/nifi/processors/pulsar/pubsub/async/TestAsyncConsumePulsarRecord.java (hunk 3 above)
Comment (z-kovacs) [reply nifi-pulsar-processors/src/test/java/org/apache/nifi/processors/pulsar/pubsub/async/TestAsyncConsumePulsarRecord.java:262]: I mean these are the exact values for the fields?

File: nifi-pulsar-processors/src/test/java/org/apache/nifi/processors/pulsar/pubsub/async/TestAsyncConsumePulsarRecord.java (hunk 4 above)
Comment (cheemau) [reply nifi-pulsar-processors/src/test/java/org/apache/nifi/processors/pulsar/pubsub/async/TestAsyncConsumePulsarRecord.java:113]: everything is mocked in this context ... there is no validation to check the schema types.

File: nifi-pulsar-processors/src/test/java/org/apache/nifi/processors/pulsar/pubsub/async/TestAsyncConsumePulsarRecord.java (hunk 5 above)
Comment (cheemau) [reply nifi-pulsar-processors/src/test/java/org/apache/nifi/processors/pulsar/pubsub/async/TestAsyncConsumePulsarRecord.java:145]: same as above

**Pull request merged**